cpp/dolfinx/fem/discreteoperators.h

```diff
@@ -68,128 +66,78 @@ void assemble_discrete_gradient(
   std::shared_ptr<const mesh::Mesh> mesh = V0.mesh();
   assert(mesh);
 
-  // Check that mesh is the same for both function spaces
-  assert(V1.mesh());
-  if (mesh != V1.mesh())
-  {
-    throw std::runtime_error("Compute discrete gradient operator. Function "
-                             "spaces do not share the same mesh");
-  }
-
-  // Check that V0 is a (lowest-order) edge basis
-  mesh->topology_mutable().create_entities(1);
-  std::int64_t num_edges_global = mesh->topology().index_map(1)->size_global();
-  const std::int64_t V0dim
-      = V0.dofmap()->index_map->size_global() * V0.dofmap()->index_map_bs();
-  if (V0dim != num_edges_global)
+  std::array<std::string, 2> lagrange_identities = {"Q", "Lagrange"};
+  std::array<std::string, 3> nedelec_identities
+      = {"Nedelec 1st kind H(curl)", "RTCE", "NCE"};
+  auto e0 = V0.element();
+  if (std::string fam0 = e0->family();
+      std::find(nedelec_identities.begin(), nedelec_identities.end(), fam0)
+      == nedelec_identities.end())
   {
     throw std::runtime_error(
-        "Cannot compute discrete gradient operator. Function "
-        "spaces is not a lowest-order edge space");
+        "Output space has to be a Nedelec (first kind) function space.");
   }
-
-  // Check that V1 is a linear nodal basis
-  const std::int64_t num_vertices_global
-      = mesh->topology().index_map(0)->size_global();
-  const std::int64_t V1dim
-      = V1.dofmap()->index_map->size_global() * V1.dofmap()->index_map_bs();
-  if (V1dim != num_vertices_global)
+  auto e1 = V1.element();
+  if (std::string fam1 = e1->family();
+      std::find(lagrange_identities.begin(), lagrange_identities.end(), fam1)
+      == lagrange_identities.end())
   {
     throw std::runtime_error(
-        "Cannot compute discrete gradient operator. Function "
-        "space is not a linear nodal function space");
+        "Output space has to be a Lagrange function space.");
   }
 
-  // Build maps from entities to local dof indices
-  const dolfinx::fem::ElementDofLayout& layout0
-      = V0.dofmap()->element_dof_layout();
-  const dolfinx::fem::ElementDofLayout& layout1
-      = V1.dofmap()->element_dof_layout();
-
-  // Copy index maps from dofmaps
-  std::array<std::shared_ptr<const common::IndexMap>, 2> index_maps
-      = {{V0.dofmap()->index_map, V1.dofmap()->index_map}};
-  std::vector<std::array<std::int64_t, 2>> local_range
-      = {index_maps[0]->local_range(), index_maps[1]->local_range()};
-  assert(V0.dofmap()->index_map_bs() == V1.dofmap()->index_map_bs());
-
-  // Initialize required connectivities
+  // Tabulate Lagrange space at H(curl) interpolation points
```

Comment:
Break up steps with comments. Looks like first step is getting the interpolation points for Lagrange element.